File: internal/common/probability/probability.go
package probability

func PermuteSpots(values []string, startSize, endSize int, emit func([]string)) {
	for i := startSize; i <= endSize; i++ {
		Permute(values, i, emit)
	}
}

func Permute(values []string, size int, emit func([]string)) {
	s := make([]string, size)
	last := size - 1
	var rc func(int, int)

	rc = func(start, next int) {
		for current, value := range values {
			s[start] = value

			if start == last {
				emit(s)
			} else {
				rc(start+1, current+1)
			}
		}
	}

	rc(0, 0)
}

func PermuteInts(values []int, size int, emit func([]int)) {
	s := make([]int, size)
	last := size - 1
	var rc func(int, int)

	rc = func(start, next int) {
		for current, value := range values {
			s[start] = value

			if start == last {
				emit(s)
			} else {
				rc(start+1, current+1)
			}
		}
	}

	rc(0, 0)
}

func ComboSize(values []int, startSize int, endSize int, emit func([]int)) {
	var permuteSize func([]int, int, int)

	permuteSize = func(current []int, index int, size int) {
		if len(current) == size {
			emit(current)
			return
		}

		for i := index; i < len(values); i++ {
			current = append(current, values[i])
			permuteSize(current, i+1, size)
			current = current[:len(current)-1]
		}
	}

	for size := startSize; size <= endSize; size++ {
		permuteSize(nil, 0, size)
	}
}

func Combo(values []int, output func([]int)) {
	combo(values, output, 0)
}

func combo(values []int, output func([]int), i int) {
	if i > len(values) {
		output(values)
		return
	}

	combo(values, output, i+1)

	for j := i + 1; j < len(values); j++ {
		values[i], values[j] = values[j], values[i]
		combo(values, output, i+1)
		values[i], values[j] = values[j], values[i]
	}
}

func ComboStrings(values []string, output func([]string)) {
	comboStrings(values, output, 0)
}

func comboStrings(values []string, output func([]string), i int) {
	if i > len(values) {
		output(values)
		return
	}

	comboStrings(values, output, i+1)

	for j := i + 1; j < len(values); j++ {
		values[i], values[j] = values[j], values[i]
		comboStrings(values, output, i+1)
		values[i], values[j] = values[j], values[i]
	}
}

File: internal/js/probability.ts
export function combinations(values: string[], startSize: number, endSize: number) {
    const result: string[][] = [];

    function combo(values: string[], size: number) {
        const s = new Array(size);
        const last = size - 1;

        const recurse = function (start: number, next: number) {
            for (let current = 0; current < values.length; current++) {
                const value = values[current];
                s[start] = value;

                if (start === last) {
                    result.push([...s]);
                } else {
                    recurse(start + 1, current + 1);
                }
            }
        };

        recurse(0, 0);
    }

    for (let i = startSize; i <= endSize; i++) {
        combo(values, i);
    }

    return result;
}

/**
 * Permutations without repeating elements. Permutations are where order matters. So 'ab' is not equivalent to 'ba'.
 * @param list
 * @param min
 * @param max
 */
export function permutations<T>(list: T[], min: number = 1, max: number = list.length) {
    let result: T[][] = [];
    for (let size = min; size <= max; size++) {
        result = result.concat(permute(list, size));
    }
    return result;
}

/**
 * Take from https://stackoverflow.com/questions/59369727/permutation-with-max-length-parameter
 * @param list
 * @param size
 */
function permute<T>(list: T[], size: number = list.length) {
    const result: T[][] = [];

    if (size === 0) {
        return [[]];
    }

    for (let i = 0; i < list.length; i++) {
        const copy = [...list];
        const head = copy.splice(i, 1);
        const rest = permute(copy, size - 1);

        for (let j = 0; j < rest.length; j++) {
            const next = head.concat(rest[j]);
            result.push(next);
        }
    }

    return result;
}
